sdk/identity/Azure.Identity/src/Credentials/OnBehalfOfCredential.cs

```diff
@@ -156,6 +180,35 @@ public OnBehalfOfCredential(string tenantId, string clientId, X509Certificate2 c
             AdditionallyAllowedTenantIds = TenantIdResolver.ResolveAddionallyAllowedTenantIds(options?.AdditionallyAllowedTenants);
         }
 
+        internal OnBehalfOfCredential(
+            string tenantId,
+            string clientId,
+            Func<string> clientAssertionCallback,
+            Func<CancellationToken, Task<string>> clientAssertionCallbackAsync,
+            string userAssertion,
+            OnBehalfOfCredentialOptions options,
+            CredentialPipeline pipeline,
+            MsalConfidentialClient client)
+        {
+            Argument.AssertNotNull(clientId, nameof(clientId));
+
+            options ??= new OnBehalfOfCredentialOptions();
+            _pipeline = pipeline ?? CredentialPipeline.GetInstance(options);
+            _tenantId = Validations.ValidateTenantId(tenantId, nameof(tenantId));
+            _clientId = clientId;
+            _userAssertion = new UserAssertion(userAssertion);
+            Client = client switch
+            {
+                not null => client,
+                _ when clientAssertionCallback is not null => new MsalConfidentialClient(_pipeline, _tenantId, _clientId, clientAssertionCallback, options),
+                _ when clientAssertionCallbackAsync is not null => new MsalConfidentialClient(_pipeline, _tenantId, _clientId, clientAssertionCallbackAsync, options),
+                _ => throw new ArgumentNullException($"nameof(clientAssertionCallback)")
```

Comment:
If this is a consistent local pattern, then feel free to ignore.

The argument name won't match in all cases, for example when both `clientAssertionCallback` and `clientAssertionCallbackAsync` are null. Worth checking separately?

Reply:
If both are null, then won't only the pure default will be chosen `(throw new ArgumentNullException(..))`?

Reply:
Yep.
To clarify my question: Are we OK with preferring `nameof(clientAssertionCallback)` in the exception message over the async param, when both are null? Looks like we have an implicit bias for that parameter, given the ordering of the check.
Seems reasonable to do. Maybe add a comment to call out the implicit ordering/preference is by design, if you think it's worth it.

Or maybe we update the exception message to say both params are null (which matches the code).

Reply:
ah - sorry, I misunderstood. 

The name will always be correct relative to the public API surface as both the sync and async overloads use the same name.

Reply:
Oh that's true! That's good.